### PR TITLE
Update to Smack 4.2

### PR DESCRIPTION
**Pull request by ibauersachs**

### CODE REVIEW

File: src/main/java/org/jitsi/impl/protocol/xmpp/extensions/ConferenceIq.java [hunk 1]
```diff
@@ -198,7 +203,17 @@ public void addProperty(Property property)
      */
     public List<Property> getProperties()
     {
-        return getChildExtensionsOfType(Property.class);
+        return (List)getExtensions(
+                Property.ELEMENT_NAME,
+                ConferenceIq.NAMESPACE);
+//        List<Property> result = new ArrayList<>();
```
Comment (bbaldino) [inline src/main/java/org/jitsi/impl/protocol/xmpp/extensions/ConferenceIq.java]: should this commented-out code still be here?

File: src/main/java/org/jitsi/impl/protocol/xmpp/extensions/ConferenceIq.java [hunk 2]
```diff
@@ -297,7 +311,7 @@ public void setIdentity(String identity)
          */
         public Property()
         {
-            super(null, ELEMENT_NAME);
+            super(ConferenceIq.NAMESPACE, ELEMENT_NAME);
```
Comment (bbaldino) [inline src/main/java/org/jitsi/impl/protocol/xmpp/extensions/ConferenceIq.java]: maybe we should add a fixme in `AbstractPacketExtension` for changing the constructor to match the order of smack 4's (elementname, namespace) as opposed to the old (namespace, elementname)

File: src/main/java/org/jitsi/impl/protocol/xmpp/OpSetSimpleCapsImpl.java [hunk 3]
```diff
@@ -23,6 +23,9 @@
 import org.jitsi.protocol.xmpp.*;
 
 import org.jivesoftware.smack.*;
+import org.jivesoftware.smack.SmackException.NoResponseException;
+import org.jivesoftware.smack.SmackException.NotConnectedException;
+import org.jxmpp.jid.Jid;
```
[the remark this answers is not in the document]
Comment (ibauersachs) [reply src/main/java/org/jitsi/impl/protocol/xmpp/OpSetSimpleCapsImpl.java]: Yes, I missed doing the cleanup here.

File: src/main/java/org/jitsi/impl/protocol/xmpp/extensions/ConferenceIq.java (hunk 2 above)
Comment (ibauersachs) [reply src/main/java/org/jitsi/impl/protocol/xmpp/extensions/ConferenceIq.java]: It seemed too dangerous switching two String parameters. I'll think about it.

Comment (bbaldino) [reply src/main/java/org/jitsi/impl/protocol/xmpp/extensions/ConferenceIq.java]: yeah agree...it's quite easy to get them wrong.  don't need to fix it in this pr, but wondered if a fixme comment would be helpful so eventually we can get everything consistent with smack.

Comment (ibauersachs) [reply src/main/java/org/jitsi/impl/protocol/xmpp/extensions/ConferenceIq.java]: A feasible way might be renaming `AbstractPacketExtension` to `AbstractExtensionElement` and changing the constructor along the way.

**Comment by ibauersachs** [comment]
Updated to address the comments and merge of master.

File: src/main/java/org/jitsi/jicofo/JitsiMeetConferenceImpl.java [hunk 4]
```diff
@@ -144,11 +151,6 @@
     private final Object participantLock = new Object();
 
     /**
-     * Takes care of conference recording.
-     */
-    private JitsiMeetRecording recording;
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jicofo/JitsiMeetConferenceImpl.java]: Why was this removed?

Comment (ibauersachs) [reply src/main/java/org/jitsi/jicofo/JitsiMeetConferenceImpl.java]: It was marked as deprecated (for a long time, if I remember it correctly), so I didn't bother migrating it.

File: src/main/java/org/jitsi/jicofo/Participant.java [hunk 5]
```diff
@@ -51,9 +54,10 @@
      * @param chatRoomMember XMPP MUC chat room member which represents a
      *                       <tt>Participant</tt>.
      */
-    public static String getEndpointId(ChatRoomMember chatRoomMember)
+    public static Jid getEndpointId(XmppChatMember chatRoomMember)
     {
-        return chatRoomMember.getName(); // XMPP MUC Nickname
+        //return chatRoomMember.getName(); // XMPP MUC Nickname
+        return chatRoomMember.getJabberID();
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jicofo/Participant.java]: This looks wrong. getJabberID lacks proper documentation but looking at some logs it seems to contain a full JID. For endpoint IDs we probably want just the resource part (of the MUC JID), which is also what getName() seems to return.

File: src/main/java/org/jitsi/jicofo/Participant.java [hunk 6]
```diff
@@ -543,7 +547,7 @@ public MediaSourceGroupMap getSourceGroups()
         MediaSourceGroupMap addedGroups = (MediaSourceGroupMap) added[1];
 
         // Mark as source owner
-        String roomJid = roomMember.getContactAddress();
+        Jid roomJid = roomMember.getJabberID();
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jicofo/Participant.java]: I think getContactAddress returns the JID in the MUC, while getJabberID return the user's own JID.

File: src/main/java/org/jitsi/jicofo/JitsiMeetConferenceImpl.java (hunk 4 above)
Comment (bgrozev) [reply src/main/java/org/jitsi/jicofo/JitsiMeetConferenceImpl.java]: Ah, that's jirecon-specific. OK, thanks.

File: src/main/java/org/jitsi/jicofo/Participant.java (hunk 5 above)
Comment (bgrozev) [reply src/main/java/org/jitsi/jicofo/Participant.java]: My comment above is obsolete, I didn't see all the commits. Please disregard.

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java [hunk 7]
```diff
@@ -23,6 +23,8 @@
 import net.java.sip.communicator.service.protocol.event.*;
 import net.java.sip.communicator.util.Logger;
 
+import org.jitsi.impl.protocol.xmpp.extensions.*;
+import org.jitsi.jicofo.*;
```
Comment (bgrozev) [inline src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java]: I thought that we were intentionally avoiding depending on jicofo classes in impl.protocol.xmpp. Is there an important reason to introduce it?
cc @paweldomas

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java [hunk 8]
```diff
@@ -650,7 +650,7 @@ public XmppChatMember findChatMember(Jid mucJid)
 
         for (ChatMemberImpl member : copy)
         {
-            if (mucJid.equals(member.getJabberID()))
+            if (mucJid.toString().equals(member.getContactAddress()))
```
Comment (bgrozev) [inline src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java]: getJabberID and getContactAddress are different

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java (hunk 7 above)
Comment (paweldomas) [reply src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java]: I guess it would be better if JitsiMeetConference would subscribe to some start muted thing instead of ChatRoomImpl setting it here.

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java [hunk 9]
```diff
@@ -217,6 +236,54 @@ public void processPresence(Presence packet)
...
+                            = presence.getExtension(
+                                StartMutedPacketExtension.ELEMENT_NAME,
+                                StartMutedPacketExtension.NAMESPACE);
+
+                        if (ext != null)
+                        {
+                            boolean[] startMuted
+                                = { ext.getAudioMuted(), ext.getVideoMuted() };
+
+                            conference.setStartMuted(startMuted);
```
Comment (paweldomas) [inline src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java]: Do we really need to move that away from the MeetExtensionsHandler ?

File: src/main/java/org/jitsi/jicofo/JitsiMeetConferenceImpl.java [hunk 10]
```diff
@@ -857,9 +857,9 @@ private boolean checkAtLeastTwoParticipants()
      * @return <tt>true</tt> if given {@link ChatRoomMember} is a focus
      *         participant.
      */
-    boolean isFocusMember(ChatRoomMember member)
+    boolean isFocusMember(XmppChatMember member)
     {
-        return member.getName().equals(focusUserName);
+        return focusUserName.equals(member.getJabberID().getResourceOrEmpty());
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jicofo/JitsiMeetConferenceImpl.java]: If I am reading the code correctly, member.getName() is the resource part of the MUC JID, while getJabberID() returns the original JID (not the one in the MUC).

File: src/main/java/org/jitsi/jicofo/JitsiMeetConferenceImpl.java [hunk 11]
```diff
@@ -1061,7 +1061,7 @@ private Participant findParticipantForJingleSession(
     {
         for (Participant participant : participants)
         {
-            if (participant.getChatMember().getContactAddress().equals(
+            if (participant.getChatMember().getJabberID().equals(
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jicofo/JitsiMeetConferenceImpl.java]: getContactAddress vs getJabberID

File: src/main/java/org/jitsi/jicofo/xmpp/BaseBrewery.java [hunk 12]
```diff
@@ -233,7 +233,7 @@ else if (ChatRoomMemberPresenceChangeEvent.MEMBER_LEFT.equals(eventType)
             || ChatRoomMemberPresenceChangeEvent.MEMBER_QUIT.equals(eventType))
         {
             // Process offline status
-            BrewInstance instance = find(chatMember.getContactAddress());
+            BrewInstance instance = find(chatMember.getJabberID());
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jicofo/xmpp/BaseBrewery.java]: getContactAddress vs getJabberID

File: src/main/java/org/jitsi/jicofo/xmpp/BaseBrewery.java [hunk 13]
```diff
@@ -265,14 +265,14 @@ private void processMemberPresence(XmppChatMember member)
...
-        PacketExtension ext
+        ExtensionElement ext
             = presence.getExtension(extensionElementName, extensionNamespace);
 
         // if the extension is missing skip processing
         if (ext == null)
             return;
 
-        processInstanceStatusChanged(member.getContactAddress(), (T) ext);
+        processInstanceStatusChanged(member.getJabberID(), (T) ext);
```
Comment (bgrozev) [inline src/main/java/org/jitsi/jicofo/xmpp/BaseBrewery.java]: getContactAddress vs getJabberID

**Comment by bgrozev** [comment]
About getJabberID and getContactAddress: I just confirmed that they do return different JIDs (the real JID and the MUC JID respectively)

**Comment by ibauersachs** [comment]
Thanks Boris! I've pushed a commit that hopefully fixes all these incorrect getContactAddress/getJabberID changes.

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java (hunk 7 above)
Comment (ibauersachs) [reply src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java]: @bgrozev Well, I couldn't know about that.
@paweldomas Not sure about the best way to do that. How about letting `JitsiMeetConferenceImpl` subscribe to presence events of `ChatRoomImpl`? Or something in `ChatRoomRoleAndPresence`? I think I was considering the latter, but didn't go for it since the necessary information (i.e. the actual presence packet to access the extension) isn't present. So a proper presence listener in `ChatRoomImpl` would still be necessary.

File: src/main/java/org/jitsi/jicofo/MeetExtensionsHandler.java [hunk 14]
```diff
@@ -87,77 +83,92 @@ public void init()
...
     {
-        if (connection == null)
+        ColibriIqHandler()
         {
-            logger.error("Not initialized");
-            return;
+            super(ColibriConferenceIQ.ELEMENT_NAME,
+                ColibriConferenceIQ.NAMESPACE,
+                IQ.Type.get,
+                Mode.sync);
```
Comment (paweldomas) [inline src/main/java/org/jitsi/jicofo/MeetExtensionsHandler.java]: I don't think we want to use Mode.sync here. If you take a look at "QueuePacketProcessor" which was used here before it was putting on a separate queue to not block the main packet reader thread. I'm guessing that the "async" mode will do the same here.

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java (hunk 7 above)
Comment (paweldomas) [reply src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java]: So the problem is that there's no longer way to add PacketListener to the connection directly ? I need to take a look.

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java [hunk 15]
```diff
@@ -217,6 +236,54 @@ public void processPresence(Presence packet)
             muc.addPresenceInterceptor(presenceInterceptor);
 
             muc.create(myNickName);
+            muc.addParticipantListener(new PresenceListener()
```
Comment (paweldomas) [inline src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java]: Do you think it would make sense to move this listener to JitsiMeetConferenceImpl directly as nested class/anonymous or something ? It knows when the chat room is created and we would get rid of "jicofo" package import.

**Comment by bgrozev** [comment]
I'm getting these exceptions when jicofo tries to connect as a client: 
```
Jicofo 2017-09-19 14:54:03.445 WARNING: [573] org.jivesoftware.smack.AbstractXMPPConnection.callConnectionClosedOnErrorListener() Connection XMPPTCPConnection[not-authenticated] (0) closed with error
javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
        at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
        at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1949)
        at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:302)
        at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:296)
        at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1509)
        at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:216)
        at sun.security.ssl.Handshaker.processLoop(Handshaker.java:979)
        at sun.security.ssl.Handshaker.process_record(Handshaker.java:914)
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1062)
        at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
        at org.jivesoftware.smack.tcp.XMPPTCPConnection.proceedTLSReceived(XMPPTCPConnection.java:798)
        at org.jivesoftware.smack.tcp.XMPPTCPConnection.access$1200(XMPPTCPConnection.java:150)
        at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader.parsePackets(XMPPTCPConnection.java:1055)
        at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader.access$300(XMPPTCPConnection.java:982)
        at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader$1.run(XMPPTCPConnection.java:998)
        at java.lang.Thread.run(Thread.java:745)
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
        at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:387)
        at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:292)
        at sun.security.validator.Validator.validate(Validator.java:260)
        at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:324)
        at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:229)
        at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:124)
        at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1491)
        ... 13 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
        at sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141)
        at sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126)
        at java.security.cert.CertPathBuilder.build(CertPathBuilder.java:280)
        at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:382)
        ... 19 more
```

Any idea how to work around them (perhaps disable certificate validation in smack, just for the tests)?

**Comment by ibauersachs** [comment]
I've forwarded you a patch for the mismatching certificate.

File: src/main/java/org/jitsi/jicofo/MeetExtensionsHandler.java (hunk 14 above)
Comment (ibauersachs) [reply src/main/java/org/jitsi/jicofo/MeetExtensionsHandler.java]: I don't mind changing this to async, but I wonder why this would be necessary. This handler is basically doing nothing (check if the get IQ has a recording set, check if there's a matching conference, return the result IQ).

File: src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java (hunk 7 above)
Comment (ibauersachs) [reply src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.java]: You can still add StanzaListeners, but not for get/set IQs (well, you can in Smack 4.2, but no longer in 4.3).

**Comment by abozhang** [comment]
@bgrozev Hi, I got the same error when installing jicofo, how did you solve this issue? Thanks a lot!

aker.java:296)
        at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1509)
        at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:216)
        at sun.security.ssl.Handshaker.processLoop(Handshaker.java:979)
        at sun.security.ssl.Handshaker.process_record(Handshaker.java:914)
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1062)
        at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
        at org.jivesoftware.smack.tcp.XMPPTCPConnection.proceedTLSReceived(XMPPTCPConnection.java:798)
        at org.jivesoftware.smack.tcp.XMPPTCPConnection.access$1200(XMPPTCPConnection.java:150)
        at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader.parsePackets(XMPPTCPConnection.java:1055)
        at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader.access$300(XMPPTCPConnection.java:982)
        at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader$1.run(XMPPTCPConnection.java:998)
        at java.lang.Thread.run(Thread.java:745)
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
        at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:387)
        at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:292)
        at sun.security.validator.Validator.validate(Validator.java:260)
        at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:324)
        at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:229)
        at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:124)
        at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1491)
        ... 13 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
        at sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141)
        at sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126)
        at java.security.cert.CertPathBuilder.build(CertPathBuilder.java:280)
        at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:382)
        ... 19 more

**Comment by bgrozev** [comment]
See [here](https://github.com/jitsi/jicofo/#certificates) and open a new issue if you still have a problem (and please describe how your environment was setup).
